File: sdk/python/src/trident_indexer/retry.py
"""Retry policy for idempotent HTTP requests (GET).

Honours the ``Retry-After`` header on 429/503 responses, falling back to
exponential backoff with jitter otherwise. Pass ``False`` in place of a
:class:`RetryConfig` to disable retries for a client or a single call.
"""

from __future__ import annotations

import random
from dataclasses import dataclass
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from typing import Literal, Optional, Union


@dataclass(frozen=True)
class RetryConfig:
    """Configurable retry policy.

    Attributes:
        max_attempts: Total number of attempts, including the first.
        base_delay: Base delay in seconds used for exponential backoff.
        max_delay: Upper bound for a single computed backoff delay.
        max_total_wait: Upper bound on total time spent waiting across all
            retries (including any honoured ``Retry-After``).
        jitter: Randomize each computed delay in ``[0, delay)``.
    """

    max_attempts: int = 3
    base_delay: float = 0.1
    max_delay: float = 2.0
    max_total_wait: float = 10.0
    jitter: bool = True


DEFAULT_RETRY_CONFIG = RetryConfig()

# `None` means "no override, use the base"; `False` disables retries. `True`
# is deliberately not accepted — only `False` carries meaning here, and
# admitting `bool` left a Literal[True] that resolve_retry_config could return
# in place of a RetryConfig.
RetryOverride = Optional[Union[RetryConfig, Literal[False]]]


def resolve_retry_config(
    override: RetryOverride,
    base: RetryOverride,
) -> Optional[RetryConfig]:
    """Merge a per-call override with the client-level base config."""
    cfg = override if override is not None else base
    if cfg is False:
        return None
    if cfg is None:
        return DEFAULT_RETRY_CONFIG
    return cfg


def is_retryable_status(status: int) -> bool:
    """Only 429 (rate limited) and 503 (service unavailable) are retried."""
    return status in (429, 503)


def parse_retry_after_seconds(header_value: object) -> Optional[float]:
    """Parse a ``Retry-After`` header, either seconds or an HTTP date."""
    if not isinstance(header_value, str):
        return None
    trimmed = header_value.strip()
    if not trimmed:
        return None
    try:
        return max(0.0, float(trimmed))
    except ValueError:
        pass
    try:
        dt = parsedate_to_datetime(trimmed)
    except (TypeError, ValueError, IndexError):
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return max(0.0, (dt - datetime.now(timezone.utc)).total_seconds())


def compute_backoff_seconds(attempt: int, cfg: RetryConfig) -> float:
    """Exponential backoff with optional full jitter, capped at max_delay."""
    # `2 ** n` is typed Any (int ** int can yield float for a negative
    # exponent), which propagates through the arithmetic and makes the return
    # Any under --strict. Use float(...) to pin it.
    exp: float = cfg.base_delay * float(2 ** (attempt - 1))
    capped: float = min(exp, cfg.max_delay)
    return random.random() * capped if cfg.jitter else capped

File: sdk/rust/src/retry.rs
use std::time::Duration;

use rand::Rng;

/// Retry policy for idempotent HTTP requests (GET). Honours `Retry-After` on
/// 429/503 responses, falling back to exponential backoff with jitter
/// otherwise.
///
/// `None` in [`TridentConfig::retry`](crate::TridentConfig::retry) (or in a
/// `*_with_retry` call) disables retries entirely — the default.
#[derive(Debug, Clone)]
pub struct RetryConfig {
    /// Total number of attempts, including the first.
    pub max_attempts: u32,
    /// Base delay used for exponential backoff.
    pub base_delay: Duration,
    /// Upper bound for a single computed backoff delay.
    pub max_delay: Duration,
    /// Upper bound on total time spent waiting across all retries
    /// (including any honoured `Retry-After`).
    pub max_total_wait: Duration,
    /// Randomize each computed delay in `[0, delay]`.
    pub jitter: bool,
}

impl Default for RetryConfig {
    fn default() -> Self {
        RetryConfig {
            max_attempts: 3,
            base_delay: Duration::from_millis(100),
            max_delay: Duration::from_secs(2),
            max_total_wait: Duration::from_secs(10),
            jitter: true,
        }
    }
}

/// Only 429 (rate limited) and 503 (service unavailable) are retried.
pub(crate) fn is_retryable_status(status: u16) -> bool {
    status == 429 || status == 503
}

/// Parse a `Retry-After` header value expressed as a number of seconds (the
/// delta-seconds form from RFC 9110; the HTTP-date form is not supported).
pub(crate) fn parse_retry_after_seconds(
    header: Option<&reqwest::header::HeaderValue>,
) -> Option<Duration> {
    header
        .and_then(|v| v.to_str().ok())
        .map(str::trim)
        .and_then(|s| s.parse::<u64>().ok())
        .map(Duration::from_secs)
}

/// Exponential backoff with optional full jitter, capped at `max_delay`.
pub(crate) fn compute_backoff(attempt: u32, cfg: &RetryConfig) -> Duration {
    let shift = attempt.saturating_sub(1).min(31);
    let exp = cfg
        .base_delay
        .saturating_mul(1u32.checked_shl(shift).unwrap_or(u32::MAX));
    let capped = exp.min(cfg.max_delay);
    if !cfg.jitter {
        return capped;
    }
    let millis = capped.as_millis() as u64;
    if millis == 0 {
        return Duration::from_millis(0);
    }
    Duration::from_millis(rand::thread_rng().gen_range(0..=millis))
}
